process/block/preprocess/transactionsV2.go

```diff
@@ -0,0 +1,416 @@
+
+		numNewMiniBlocks := 0
+		if len(miniBlock.TxHashes) == 0 || firstCrossShardScCallOrSpecialTxFound {
+			numNewMiniBlocks = 1
+		}
+		numNewTxs := 1
+
+		isCrossShardScCallOrSpecialTx := receiverShardID != txs.shardCoordinator.SelfId() &&
+			(isReceiverSmartContractAddress || len(tx.RcvUserName) > 0)
+		if isCrossShardScCallOrSpecialTx {
+			if !firstCrossShardScCallOrSpecialTxFound {
+				numNewMiniBlocks++
+			}
+			numNewTxs += core.AdditionalScrForEachScCallOrSpecialTx
+		}
+
+		if isMaxBlockSizeReached(numNewMiniBlocks, numNewTxs) {
+			log.Debug("max txs accepted in one block is reached",
+				"num txs added", numTxsAdded,
+				"total txs", len(sortedTxs))
+			break
+		}
+
+		if isShardStuck != nil && isShardStuck(receiverShardID) {
+			log.Trace("shard is stuck", "shard", receiverShardID)
+			continue
+		}
+
+		if len(senderAddressToSkip) > 0 {
+			if bytes.Equal(senderAddressToSkip, tx.GetSndAddr()) {
+				numTxsSkipped++
+				continue
+			}
+		}
+
+		txMaxTotalCost := big.NewInt(0)
+		isAddressSet := txs.balanceComputation.IsAddressSet(tx.GetSndAddr())
+		if isAddressSet {
+			txMaxTotalCost = txs.getTxMaxTotalCost(tx)
+		}
+
+		if isAddressSet {
+			addressHasEnoughBalance := txs.balanceComputation.AddressHasEnoughBalance(tx.GetSndAddr(), txMaxTotalCost)
+			if !addressHasEnoughBalance {
+				numTxsWithInitialBalanceConsumed++
+				continue
+			}
+		}
+
+		snapshot := txs.accounts.JournalLen()
+
+		gasConsumedByMiniBlockInReceiverShard := mapGasConsumedByMiniBlockInReceiverShard[receiverShardID]
+		oldGasConsumedByMiniBlocksInSenderShard := gasConsumedByMiniBlocksInSenderShard
+		oldGasConsumedByMiniBlockInReceiverShard := gasConsumedByMiniBlockInReceiverShard
+		oldTotalGasConsumedInSelfShard := totalGasConsumedInSelfShard
+		startTime := time.Now()
+		err := txs.computeGasConsumed(
+			senderShardID,
+			receiverShardID,
+			tx,
+			txHash,
+			&gasConsumedByMiniBlocksInSenderShard,
+			&gasConsumedByMiniBlockInReceiverShard,
+			&totalGasConsumedInSelfShard)
+		elapsedTime := time.Since(startTime)
+		totalTimeUsedForComputeGasConsumed += elapsedTime
+		if err != nil {
+			log.Trace("createAndProcessMiniBlocksFromMeV2.computeGasConsumed", "error", err)
+			continue
+		}
+
+		mapGasConsumedByMiniBlockInReceiverShard[receiverShardID] = gasConsumedByMiniBlockInReceiverShard
+
+		// execute transaction to change the trie root hash
+		startTime = time.Now()
+		err = txs.processAndRemoveBadTransaction(
+			txHash,
+			tx,
+			senderShardID,
+			receiverShardID,
+		)
+		elapsedTime = time.Since(startTime)
+		totalTimeUsedForProcesss += elapsedTime
+
+		txs.mutAccountsInfo.Lock()
+		txs.accountsInfo[string(tx.GetSndAddr())] = &txShardInfo{senderShardID: senderShardID, receiverShardID: receiverShardID}
+		txs.mutAccountsInfo.Unlock()
+
+		if err != nil && !errors.Is(err, process.ErrFailedTransaction) {
+			if errors.Is(err, process.ErrHigherNonceInTransaction) {
```

Comment:
take out this to another function. this may be the same as with V1

Reply:
Done, for V1 I will let as it is